### DISCUSSION TITLE
Unable to complete a digital certificate (.p12) authentication

### DISCUSSION THREAD

**Discussion by dantard** [qutebrowser/qutebrowser#7556] [Q&A]
**Version info**:

qutebrowser v2.5.2
Git commit: e753974cb on master (2023-01-02 10:24:08 +0100)
Backend: QtWebEngine 5.15.2, based on Chromium 83.0.4103.122
Qt: 5.15.2

CPython: 3.8.10
PyQt: 5.15.7

sip: 6.6.2
colorama: 0.4.6
jinja2: 3.1.2
pygments: 2.14.0
yaml: 6.0
adblock: 0.6.0
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebEngine: 5.15.6
PyQt5.QtWebKitWidgets: no
objc: no
pdf.js: no
sqlite: 3.33.0
QtNetwork SSL: OpenSSL 1.1.1f  31 Mar 2020

Style: QFusionStyle
Platform plugin: xcb
OpenGL: Intel, 4.6 (Compatibility Profile) Mesa 21.2.6
Platform: Linux-5.15.0-41-generic-x86_64-with-glibc2.29, 64bit
Linux distribution: Ubuntu 20.04.4 LTS (ubuntu)
Frozen: False
Imported from /home/danilo/tmp/qutebrowser/qutebrowser
Using Python from /home/danilo/tmp/qutebrowser/.venv/bin/python3
Qt library executable path: /home/danilo/tmp/qutebrowser/.venv/lib/python3.8/site-packages/PyQt5/Qt5/libexec, data path: /home/danilo/tmp/qutebrowser/.venv/lib/python3.8/site-packages/PyQt5/Qt5

Paths:
cache: /home/danilo/.cache/qutebrowser
config: /home/danilo/.config/qutebrowser
data: /home/danilo/.local/share/qutebrowser
runtime: /run/user/1001/qutebrowser
system data: /usr/share/qutebrowser

Autoconfig loaded: yes
Config.py: no config.py was loaded
Uptime: 0:00:05

**Does the bug happen if you start with `--temp-basedir`?**:
Yes
**Description**
The browser is unable to complete an authentication process based on a .p12 digital certificate. The certificate is correctly installed and when accessing a web that has this mean of authentication, the browser does raise a popup asking if the certificate should be presented to the website but after pressing 'y' the website complains that "no certificate has been provided" (I mean, it does not complain about not being valid or expired). I tested it in several different websites and the the result is always the same.

**How to reproduce**
It is possible to reproduce the error visiting, for example: https://aesso.unizar.es/
and choosing the "Certificado digital/DNI" button (left one).

**Comment by The-Compiler** [comment]
How do you determine that it's correctly installed? Does it show up with `certutil -d "sql:${HOME}/.pki/nssdb" -L`? Can you try if it works in Chromium?

**Comment by mschilli87** [comment]
@dantard: If I click on the DNI button I don't even get a popup. :shrug:

**Comment by dantard** [comment]
> How do you determine that it's correctly installed? Does it show up with `certutil -d "sql:${HOME}/.pki/nssdb" -L`? Can you try if it works in Chromium?

Hi,

Yes, it shows up with certutils, and Yes, it works with chromium. However, I've been testing the issue with a C++ minimal test program (because I thought the problem could be related with the Python wrapper) and it doesn't work either despite the certificate is present and is correctly selected in the same way as in qutebrowser. No error messages or warning are generated either.

**Comment by dantard** [comment]
> @dantard: If I click on the DNI button I don't even get a popup. shrug

That's weird, it should popup if you have a certificate installed in the nssdb.
Can you try clicking the right button at  https://sso.us.es/CAS/index.php/login?service=https%3A%2F%2Fbareprof.us.es%2Findex.php%2Fcas%2Fauth%2Flogin 

Thanks.

**Comment by mschilli87** [comment]
@dantard: Sorry for not getting back to you. I won't have to to investigate for another week. If I didn't reply by the end of next week, feel free to ping me here. :sweat_smile:

**Comment by mschilli87** [comment]
Finally got around testing this and now that I have properly installed my cert (as per `certutil`) I can *not* replicate what @dantard describes.

*edit*: I originally wrote I can replicate but I was just not patient enough for the webpage @dantard linked above. :sweat_smile: 
So no problem on my end.

**Reply by mschilli87** [reply]
@dantard: For you to have a starting point for further investigation, here are my `:version` information:

```
qutebrowser v2.5.2
Git commit: 
Backend: QtWebEngine 5.15.7, based on Chromium 87.0.4280.144
Qt: 5.15.7

CPython: 3.10.9
PyQt: 5.15.7

sip: 6.6.2
colorama: 0.4.6
jinja2: 3.1.2
pygments: 2.13.0
yaml: 6.0
adblock: 0.6.0
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebEngine: 5.15.6
PyQt5.QtWebKitWidgets: no
pdf.js: 2.14.305 (/usr/share/pdf.js/build/pdf.js)
sqlite: 3.40.1
QtNetwork SSL: OpenSSL 1.1.1s  1 Nov 2022
```

**Comment by The-Compiler** [comment]
Thanks @mschilli87 for testing! I'm converting this to a discussion then, as this seems like an issue on @dantard's end - I don't see anything that qutebrowser could fix here, if the certificate prompt works, qutebrowser's part is pretty much working. Everything after that happens in QtWebEngine/Chromium.

**Reply by dantard** [reply]
Sure, thank you.
Yes, I believe the "problem" (too strict policy?) is inside QtWebEngine in fact. The surprising bit is that everything works nicely in chromium.

**Comment by port19x** [comment]
I found [this](https://wiki.archlinux.org/title/User:Grawity/Adding_a_trusted_CA_certificate) arch wiki page particularly helpful for setting additional certs. @dantard

**Reply by dantard** [reply]
Thank you very much but the problem is not about the certificate that is valid and recognized by qutebrowse. As I mentioned above, I believe the problem is in the Qt library.
